src/queryBuilder.ts

```diff
@@ -0,0 +1,152 @@
+                // This is an error condition. FHIR standard expects date based search only to have prefix.
+                // If prefixes (eq,gt etc) added in any other place of search value, it will be treated as part of search value.
+                return { prefix: undefined, value };
+            }
+            return { prefix: formatPrefix(ALLOWED_PREFIXES[i]), value: searchValue[1] };
+        }
+    }
+    return { prefix: undefined, value };
+}
+
+function dateBasedSearch(searchParameter: string, searchvalue: string): any {
+    const search = splitPrefixAndValue(searchvalue);
+    let query = {};
+
+    // No Prefix is available for dates
+    if (search.prefix === undefined || search.prefix === PREFIXES.EQUAL) {
+        query = {
+            range: {
+                [searchParameter]: {
+                    gte: search.value,
+                    lte: search.value,
+                },
+            },
+        };
+        return { typeOfQuery: TYPE_OF_QUERY.FILTER, query };
+    }
+
+    // NOT EQUAL Prefix
+    if (search.prefix === PREFIXES.NOT_EQUAL) {
+        query = {
+            range: {
+                [searchParameter]: {
+                    gte: search.value,
+                    lte: search.value,
+                },
+            },
+        };
+        return { typeOfQuery: TYPE_OF_QUERY.MUST_NOT, query };
+    }
+
+    // Need to implement approximations
+
+    // All other prefixes
+    query = {
+        range: {
+            [searchParameter]: {
+                [search.prefix]: search.value,
+            },
+        },
+    };
+    return { typeOfQuery: TYPE_OF_QUERY.FILTER, query };
+}
+
+// Split the query based on search field data types
+function splitQuery(field: string, searchParameter: string, value: string) {
+    if (DATE_FIELDS.includes(searchParameter)) return dateBasedSearch(searchParameter, value);
+    //  Use for all other search parameters.
+    // TODO: Need to refine this search parameter
+    const query = {
+        query_string: {
+            fields: [field],
+            query: value,
+            default_operator: 'AND',
+            lenient: true,
+        },
+    };
+    return { typeOfQuery: TYPE_OF_QUERY.MUST, query };
+}
+
+/**
+ * @param Pass Query strings  {"parameter1": "prefixvalue1", "parameter2": "prefixvalue2"}
+ * ex - {"birthDate": "eq2020-12-01"}
+ * @param filterRulesForActiveResources - If you are storing both History and Search resources
+ * in your elastic search you can filter out your History elements by supplying a filter argument like:
+ * [{ match: { documentStatus: 'AVAILABLE' }}]
+ * @param Returns built ES query
+ */
+
+export function buildQuery(queryParams: any, filterRulesForActiveResources: any): any {
+    const must: any[] = [];
+    const filter: any[] = [];
+    const mustNot: any[] = [];
+    // Filter based on the user request
+    if (filterRulesForActiveResources.length > 0) filter.push(filterRulesForActiveResources);
```

Comment:
There's still an unintended change on the snapshots. the `filter` is now an array of arrays of query objects instead of an array of  query objects. The query itself is valid, but it's better to avoid an unnecessary nested array.

I believe that the following change + regenerating the snapshots should fix it:

```suggestion
    if (filterRulesForActiveResources.length > 0) filter.push(...filterRulesForActiveResources);
```

Reply:
Thanks Carvantes. Fixed the issue. Happy Christmas !!!